napari/_qt/threading.py

```diff
@@ -0,0 +1,563 @@
+                if self._resume_requested:
+                    self._paused = False
+                    self._resume_requested = False
+                    self.resumed.emit()
+                else:
+                    time.sleep(self._pause_interval)
+                    continue
+            elif self._pause_requested:
+                self._paused = True
+                self._pause_requested = False
+                self.paused.emit()
+                continue
+            try:
+                self.pre_yield_hook()
+                self.yielded.emit(self._gen.send(self._next_value()))
+                self.post_yield_hook()
+            except StopIteration as exc:
+                self.returned.emit(exc.value)
+                break
+
+    def send(self, value: Any):
+        """Send a value into the function (if a generator was used)."""
+        self._incoming_value = value
+
+    def _next_value(self) -> Any:
+        out = None
+        if self._incoming_value is not None:
+            out = self._incoming_value
+            self._incoming_value = None
+        return out
+
+    @property
+    def is_paused(self) -> bool:
+        """Whether the worker is currently paused."""
+        return self._paused
+
+    def toggle_pause(self) -> None:
+        """Send a request to pause the worker if playing or resumed if paused.
+        """
+        if self.is_paused:
+            self._resume_requested = True
+        else:
+            self._pause_requested = True
+
+    def pre_yield_hook(self):
+        """Hook for subclasses. Called just before yielding from generator"""
+        pass
+
+    def post_yield_hook(self):
+        """Hook for subclasses. Called just after yielding from generator"""
+        pass
+
+
+class ProgressWorker(Worker):
+    """A Worker that emits a progress update on each yield.
+
+    See Worker docstring for details.  This simply counts the number of
+    iterations, and emits a progress signal on every iteration.
+    """
+
+    # emitted on yield ONLY if the function was a generator AND it yielded
+    # dict in the first yield with a key "__len__" that defines the number of
+    # total yields in the generator.  Will emit an integer between 0 - 100
+    # every time a value is yielded
+    progress = Signal(int)
+
+    def __init__(self, *args, **kwargs):
+        self._counter = 0
+        self._length = None
+        super().__init__(*args, **kwargs)
+
+    def parse_first_yield(self, result):
+        self._length = result.pop("__len__", None)
+        if self._length is not None:
+            if not isinstance(self._length, int) and self._length > 0:
+                raise ValueError(
+                    "If providing __len__, it must be a positive int"
+                )
+        super().parse_first_yield(result)
+
+    def pre_yield_hook(self):
+        self._counter += 1
+        if self._length:
+            self.progress.emit(round(100 * (self._counter - 1) / self._length))
```

Comment:
it needs a len and an iter, yes. I feel like `__iter__` could be the same as pre_yield_hook, more or less?

Reply:
I played with this for a while actually... and while i got it to "work" (in the sense that tqdm worked), I never got around having the iterator block in the main thread again when used with tqm.  (which kinda makes sense... since it's the main thread calling "next").  Maybe we can leave this elaboration for a later PR?

Reply:
Yeah, sure, I just think then maybe we could have a len and a counter and leave the calculation to the caller?

I'd also emit first and increment the counter second so that you don't need the -1?